compiler/next/lib/AST/Builder.cpp

```diff
@@ -0,0 +1,190 @@
+  } else {
+    return std::string(moduleName);
+  }
+}
+
+owned<Builder> Builder::build(Context* context, const char* filepath) {
+  // compute the basename of filename to get the inferred module name
+  std::string modname = filenameToModulename(filepath);
+  auto uniqueModname = UniqueString::build(context, modname);
+  auto uniqueFilename = UniqueString::build(context, filepath);
+  auto b = new Builder(context, uniqueFilename, uniqueModname);
+  return toOwned(b);
+}
+
+void Builder::addToplevelExpr(owned<Expr> e) {
+  this->topLevelExprs_.push_back(std::move(e));
+}
+
+void Builder::addError(ErrorMessage e) {
+  this->errors_.push_back(std::move(e));
+}
+
+void Builder::noteLocation(BaseAST* ast, Location loc) {
+  this->locations_.push_back(std::make_pair(ast->id(), loc));
+}
+
+Builder::Result Builder::result() {
+  UniqueString inferredName = this->createImplicitModuleIfNeeded();
+  this->assignIDs(inferredName);
+
+  // Performance: We could consider copying all of these AST
+  // nodes to a newly allocated buffer big enough to hold them
+  // all contiguously. The reason to do so would be to ensure
+  // that a postorder traversal of the AST has good data locality
+  // (i.e. good cache behavior).
+
+  Builder::Result ret;
+  ret.topLevelExprs.swap(topLevelExprs_);
+  ret.errors.swap(errors_);
+  ret.locations.swap(locations_);
+  return ret;
+}
+
+// Returns the name of the implicit module, or "" if there is none
+// If the implicit module is needed, moves the statements in to it.
+UniqueString Builder::createImplicitModuleIfNeeded() {
+  bool containsOnlyModules = true;
+  bool containsAnyModules = false;
+  for (auto const& ownedExpr: topLevelExprs_) {
+    if (ownedExpr->isModuleDecl()) {
+      containsAnyModules = true;
+    } else {
+      containsOnlyModules = false;
+    }
+  }
+  if (containsAnyModules && containsOnlyModules) {
+    UniqueString empty;
+    return empty;
+  } else {
+    // create a new module containing all of the statements
+    ASTList stmts;
+    stmts.swap(topLevelExprs_);
+    auto implicitModule = ModuleDecl::build(this, Location(filepath_),
+                                            inferredModuleName_,
+                                            Symbol::VISIBILITY_DEFAULT,
+                                            Module::IMPLICIT,
+                                            std::move(stmts));
+    topLevelExprs_.push_back(std::move(implicitModule));
+    // return the name of the module
+    return inferredModuleName_;
+  }
+}
+
+void Builder::assignIDs(UniqueString inferredModule) {
+  pathVecT path;
+  declaredHereT decl;
+
+  if (!inferredModule.isEmpty())
+    path.push_back(std::make_pair(inferredModule, 0));
+
+  for (auto const& ownedExpr: topLevelExprs_) {
+    const BaseAST* ast = ownedExpr.get();
+    assignIDs((BaseAST*)ast, path, decl); 
+  }
+}
+
+void Builder::assignIDs(BaseAST* ast, pathVecT& path, declaredHereT& decl) {
```

Comment:
One thing that occurs to me - with creating the ids later when we know the stable order (if I'm understanding correctly), we're not going to be able to do "break-on-id" any more and have compiler debugging interrupt when the node with the id number is created.  And it sounded like the memory address for it was also created later (since it's stable between runs even if code is added later).  What would you envision doing to break when a file is parsed if you're trying to look at the earliest starting state of a particular node?

Reply:
That's an interesting question. I think we could probably shift to, for debugging parsing issues specifically, breaking on the creation of an AST node at a particular line/column. However I have not tried to add any debugging help yet.

Reply:
> with creating the ids later when we know the stable order (if I'm understanding correctly), we're not going to be able to do "break-on-id" any more and have compiler debugging interrupt when the node with the id number is created

This may occasionally be a useful feature to have. However, this solution sounds reasonable:

> we could probably shift to, for debugging parsing issues specifically, breaking on the creation of an AST node at a particular line/column

Perhaps with the ability to specialize on the kind of node that is created.

We could also add a "break-on-id" upon ASSIGNING an ID, however that does not sound very useful to me. "Aha, this ID was created during parsing!" [When else? duh] :)